pkg/controller/direct/dataform/repository_controller.go

```diff
@@ -0,0 +1,253 @@
+	log.V(2).Info("creating object", "u", u)
+
+	projectID := a.projectID
+	if projectID == "" {
+		return fmt.Errorf("project is empty")
+	}
+	if a.resourceID == "" {
+		return fmt.Errorf("resourceID is empty")
+	}
+
+	desired := a.desired.DeepCopy()
+	mapCtx := &direct.MapContext{}
+	resource := DataformRepositorySpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return fmt.Errorf("converting DataformRepository spec to api: %w", mapCtx.Err())
+	}
+
+	req := &dataformpb.CreateRepositoryRequest{
+		Parent:       a.getParent(),
+		Repository:   resource,
+		RepositoryId: a.resourceID,
+	}
+	_, err := a.gcpClient.CreateRepository(ctx, req)
+	if err != nil {
+		return fmt.Errorf("DataformRepository %s creation failed: %w", resource.Name, err)
+	}
+
+	status := &krm.DataformRepositoryStatus{}
+
+	// TODO(acpana): add observed state
+	// status.ObservedState.CreateTime = ToOpenAPIDateTime(updated.GetCreateTime())
+	// status.ObservedState.UpdateTime = ToOpenAPIDateTime(updated.GetUpdateTime())
+	return setStatus(u, status)
+}
+
+func (a *Adapter) Update(ctx context.Context, u *unstructured.Unstructured) error {
+
+	updateMask := &fieldmaskpb.FieldMask{}
+	// TODO(acpana): handle updates
+
+	desired := a.desired.DeepCopy()
+	mapCtx := &direct.MapContext{}
+	resource := DataformRepositorySpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return fmt.Errorf("converting DataformRepository spec to api: %w", mapCtx.Err())
+	}
+
+	req := &dataformpb.UpdateRepositoryRequest{UpdateMask: updateMask, Repository: resource}
+	_, err := a.gcpClient.UpdateRepository(ctx, req)
+	if err != nil {
+		return fmt.Errorf("DataformRepository %s update failed: %w", resource.Name, err)
+	}
+
+	status := &krm.DataformRepositoryStatus{}
+
+	// TODO(acpana): add observed state
+	// status.ObservedState.CreateTime = ToOpenAPIDateTime(updated.GetCreateTime())
+	// status.ObservedState.UpdateTime = ToOpenAPIDateTime(updated.GetUpdateTime())
+	return setStatus(u, status)
+}
+
+func (a *Adapter) Export(ctx context.Context) (*unstructured.Unstructured, error) {
+	// TODO(kcc)
+	return nil, fmt.Errorf("Export not supported for DataformRepository yet")
+}
+
+// Delete implements the Adapter interface.
+func (a *Adapter) Delete(ctx context.Context) (bool, error) {
+	if a.resourceID == "" {
+		return false, nil
+	}
+
+	req := &dataformpb.DeleteRepositoryRequest{Name: a.fullyQualifiedName()}
+	if err := a.gcpClient.DeleteRepository(ctx, req); err != nil {
+		return false, fmt.Errorf("deleting DataformRepository %s: %w", a.fullyQualifiedName(), err)
+	}
+
+	return true, nil
+}
+
+func (a *Adapter) fullyQualifiedName() string {
+	return fmt.Sprintf("projects/%s/locations/%s/repositories/%s", a.projectID, a.location, a.resourceID)
+}
+
+func (a *Adapter) getParent() string {
+	return fmt.Sprintf("projects/%s/locations/%s", a.projectID, a.location)
+}
+
+func setStatus(u *unstructured.Unstructured, typedStatus any) error {
```

Comment:
If it's alpha (and therefore uses observedState) we might actually use a different helper `setObservedState`; it's a much simpler function to write

Reply:
Is it ok if I follow up here? There's a few changes to the API that are needed for the beta promotion and there's no ObservedState yet in the Status. I can add it with the other API changes.

Reply:
Of course!